src/core/server/elasticsearch/client/configure_client.test.ts

```diff
@@ -0,0 +1,204 @@
+}): RequestEvent<T> => {
+  return {
+    body,
+    statusCode,
+    headers,
+    warnings,
+    meta: {
+      request: {
+        params: params!,
+      } as any,
+    } as any,
+  };
+};
+
+describe('configureClient', () => {
+  let logger: ReturnType<typeof loggingSystemMock.createLogger>;
+  let config: ElasticsearchClientConfig;
+
+  beforeEach(() => {
+    logger = loggingSystemMock.createLogger();
+    config = createFakeConfig();
+    parseClientOptionsMock.mockReturnValue({});
+    ClientMock.mockImplementation(() => createFakeClient());
+  });
+
+  afterEach(() => {
+    parseClientOptionsMock.mockReset();
+    ClientMock.mockReset();
+  });
+
+  it('calls `parseClientOptions` with the correct parameters', () => {
+    configureClient(config, { logger, scoped: false });
+
+    expect(parseClientOptionsMock).toHaveBeenCalledTimes(1);
+    expect(parseClientOptionsMock).toHaveBeenCalledWith(config, false);
+
+    parseClientOptionsMock.mockClear();
+
+    configureClient(config, { logger, scoped: true });
+
+    expect(parseClientOptionsMock).toHaveBeenCalledTimes(1);
+    expect(parseClientOptionsMock).toHaveBeenCalledWith(config, true);
+  });
+
+  it('constructs a client using the options returned by `parseClientOptions`', () => {
+    const parsedOptions = {
+      nodes: ['http://localhost'],
+    };
+    parseClientOptionsMock.mockReturnValue(parsedOptions);
+
+    const client = configureClient(config, { logger, scoped: false });
+
+    expect(ClientMock).toHaveBeenCalledTimes(1);
+    expect(ClientMock).toHaveBeenCalledWith(parsedOptions);
+    expect(client).toBe(ClientMock.mock.results[0].value);
+  });
+
+  it('listens to client on `response` events', () => {
+    const client = configureClient(config, { logger, scoped: false });
+
+    expect(client.on).toHaveBeenCalledTimes(1);
+    expect(client.on).toHaveBeenCalledWith('response', expect.any(Function));
+  });
+
+  describe('Client logging', () => {
+    it('logs error when the client emits an error', () => {
+      const client = configureClient(config, { logger, scoped: false });
+
+      const response = createApiResponse({
+        body: {
+          error: {
+            type: 'error message',
+          },
+        },
+      });
+      client.emit('response', new errors.ResponseError(response), null);
+      client.emit('response', new Error('some error'), null);
+
+      expect(loggingSystemMock.collect(logger).error).toMatchInlineSnapshot(`
+        Array [
+          Array [
+            "ResponseError: error message",
+          ],
+          Array [
+            "Error: some error",
+          ],
+        ]
+      `);
+    });
+
+    it('logs each queries if `logQueries` is true', () => {
+      const client = configureClient(
+        createFakeConfig({
+          logQueries: true,
+        }),
+        { logger, scoped: false }
+      );
+
+      const response = createApiResponse({
+        body: {},
+        statusCode: 200,
```

Comment:
I'm not really sure of what is passed to the `response` event handler in case of failed requests and/or errors tbh

 @delvedor maybe you can help us here? https://github.com/elastic/kibana/blob/b265759a323a945f1d633e7b4b9d7baaf1971f8e/src/core/server/elasticsearch/client/configure_client.ts#L37-L42

are there some cases where the handler will receive both the `error` and `event` parameters? What is the correct way to log failed requests while still properly logging other kind of errors (other than `ResponseError` that are the one 'returned' for a failed request I guess)

Reply:
Any error (Response, Connection, Timeout, Deserialization...) will be passed in the error argument.
The `event` value will always be populated, but in case of error you can find the same information in the error object, for example:
```js
client.on('response', (err, event) => {
  if (err) {
    console.error(err)
    console.error(err.meta) // same content as event
  } else {
    console.log(event)
  }
})
```

Be aware that not all errors have the same keys, as you can see [here](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-usage.html#_error_handling).